src/components/application_manager/src/state_controller_impl.cc

```diff
@@ -873,11 +909,20 @@ void StateControllerImpl::OnStateChanged(ApplicationSharedPtr app,
     return;
   }
 
-  auto notification =
-      MessageHelper::CreateHMIStatusNotification(app, window_id);
-  app_mngr_.GetRPCService().ManageMobileCommand(notification,
-                                                commands::Command::SOURCE_SDL);
-
+  if (app->is_ready()) {
+    SDL_LOG_DEBUG("Sending OnHMIStatus to application " << app->app_id());
+    auto notification =
+        MessageHelper::CreateHMIStatusNotification(app, window_id);
+    app_mngr_.GetRPCService().ManageMobileCommand(
+        notification, commands::Command::SOURCE_SDL);
+  } else {
+    SDL_LOG_DEBUG(
+        "Application "
+        << app->app_id()
+        << " not ready to receive OnHMIStatus. Delaying notification");
+    pending_hmistatus_notification_apps_.insert(app->app_id());
```

Comment:
There should probably be a lock used around pending_hmistatus_notification_apps_ usage